### ISSUE TITLE
How could the network train batch norm?

### ISSUE THREAD

**Issue by John1231983** [zhengyang-wang/Deeplab-v2--ResNet-101--Tensorflow#17]
This is not an issue. I just want to extend the training process by train with batch norm. As you mentioned, the training BN may not good when the batch size is small. However, I am running on a powerful computer so I think it can train with the batch size of 16. After completed, these BN will be frozen, and then the network trains with small batch size and learning rate.

As you mentioned in the README
>Example: If you have a batch normalization layer in the decoder, you should use 
outputs = self._batch_norm(inputs, name='g_bn1', is_training=self.phase, activation_fn=tf.nn.relu, trainable=True)

To train with BN, I will set **`is_training flag` to `True`** and  **`trainable=True,`** in the network.py. Is that all? Do I need change something in the model.py in the lines
```
 restore_var = [v for v in tf.global_variables() if 'fc' not in v.name]
# Trainable Variables
all_trainable = tf.trainable_variables()
# Fine-tune part
encoder_trainable = [v for v in all_trainable if 'fc' not in v.name] # lr * 1.0
# Decoder part
decoder_trainable = [v for v in all_trainable if 'fc' in v.name]
		....
decoder_w_trainable = [v for v in decoder_trainable if 'weights' in v.name or 'gamma' in v.name] # lr * 10 

decoder_b_trainable = [v for v in decoder_trainable if 'biases' in v.name or 'beta' in v.name] # lr * 20.0
```

This is my completed code for train BN in the decoder

```
o=self._conv2d_bn(x, 1, 256, 1, name='fc1', biased=True)
o_bn = self._batch_norm(o, name='fc1_bn', is_training=self.phase, trainable=self.phase,activation_fn=tf.nn.relu)
```

Thanks so much

**Comment by zhengyang-wang** [comment]
I think you are right. 
First, change trainable to True. (You don't have to set it to False when testing. It's different from is_training.)
Second, set is_training True for training and False for testing (by using phase). 
Third, add gamma and beta to trainable variables. (Already there.)

And two suggestions:
1. Even 16 is not a large batch size. But you can test.
2. When you use batch_norm after convolution, the bias of convolution should be set to False. (Check the paper of batch norm.)

Thanks.

**Comment by John1231983** [comment]
Thank you so much for your suggestions
